static/settings.ts

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) 2016, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import $ from 'jquery';
+
+import { SiteSettings } from './settings.interfaces';
+import { options } from './options';
+import * as colour from './colour';
+import * as local from './local';
+
+const themes = require('./themes').themes
+
+class ISetting {
+    elem: JQuery;
+    name: string;
+
+    constructor(elem, name) {
+        this.elem = elem;
+        this.name = name;
+    }
+
+    getUi(): any {
+        return this.elem.val();
+    }
+
+    putUi(value): void {
+        this.elem.val(value);
+    }
+}
+
+class Checkbox extends ISetting {
+    override getUi(): any {
+        return !!this.elem.prop('checked');
+    }
+
+    override putUi(value) {
+        this.elem.prop('checked', !!value);
+    }
+}
+
+class Select extends ISetting {
+    constructor(elem, name, populate) {
+        super(elem, name);
+
+        elem.empty();
+        for (let e of populate) {
+            elem.append($(`<option value="${e.label}">${e.desc}</option>`))
+        }
+    }
+
+    override putUi(value) {
+        this.elem.val(value.toString());
+    }
+}
+
+class Slider extends ISetting {
+    constructor(elem, name, sliderSettings) {
```

Comment:
Could type arguments as

```suggestion
    constructor(elem: JQuery, name: string, sliderSettings: Record<'max' | 'min' | 'step', number>) {
```